server/src/main/java/org/candlepin/guice/CandlepinContextListener.java

```diff
@@ -206,17 +214,45 @@ public void withInjector(Injector injector) {
 
     @Override
     public void contextDestroyed(ServletContextEvent event) {
-        super.contextDestroyed(event);
+
         if (config.getBoolean(ACTIVEMQ_ENABLED)) {
             activeMQContextListener.contextDestroyed();
+            injector.getInstance(ActiveMQStatusMonitor.class).stop();
+            ActiveMQClient.clearThreadPools();
         }
+
         pinsetterListener.contextDestroyed();
         loggerListener.contextDestroyed();
 
         // if amqp is enabled, close all connections.
         if (config.getBoolean(ConfigProperties.AMQP_INTEGRATION_ENABLED)) {
             Util.closeSafely(injector.getInstance(AMQPBusPublisher.class), "AMQPBusPublisher");
         }
+
+        injector.getInstance(PersistService.class).stop();
+        super.contextDestroyed(event);
+
+        // deregister jdbc driver to avoid warning in tomcat shutdown log
+        Enumeration<Driver> drivers = DriverManager.getDrivers();
+        while (drivers.hasMoreElements()) {
+            Driver driver = drivers.nextElement();
+            try {
+                DriverManager.deregisterDriver(driver);
+            }
+            catch (SQLException e) {
+                log.info(String.format("Failed to de-registering driver %s", driver), e);
+            }
+        }
+
+        InternalThreadLocalMap.destroy();
+
+        try {
+            Thread.sleep(1000L);
```

Comment:
I added this to complete driver de-registration process otherwise following warning messages related to connections/pools in the logs. 

“The web application [/candlepin] appears to have started a thread named [Resource Destroyer in BasicResourcePool.close()] but has failed to stop it. This is very likely to create a memory leak”

“The web application [/candlepin] appears to have started a thread named [C3P0PooledConnectionPoolManager[identityToken->z8kflta61sqd6ytby7ezy|2350d322]-AdminTaskTimer] but has failed to stop it. This is very likely to create a memory leak.”

Reply:
Is there a way to wait for the task to complete more directly (like a thread.join style of wait) rather than a short 1s sleep? It'd be nice to have something that is more direct than something like this which could introduce a race condition. Granted, it's just warning output, but we should avoid it if we can.

Reply:
@Ceiu 
Thread.currentThread().join() is a option. If I used this, there are no warning messages in the logs including thread-local one. Also, no need to code to de-register deriver. 

However, with this, the tomcat stop will take some time.

Reply:
Bit of a miscommunication here. I was using Thread.join as an example of the behavior that would be preferable to an unlinked wait. Thread.currentThread().join() shouldn't do anything positive for us here. In fact, if it works at all, it will deadlock (tells the current thread to wait until it finishes waiting...). That might explain the extra 90s of waiting: the system is completely hung during shutdown until Tomcat forcefully kills the process.

If there's nothing better in the C3P0 API to let us wait on its shutdown routines, then we can leave it as a 1s delay, but it should probably have a comment attached to it so when someone stumbles upon this a year or more from now, they'll know why there's a random sleep in the shutdown code.

Reply:
@Ceiu  
After fixing merge conflicts and changing a few cleanup code sequences, there is no warning related to pool/connection without adding sleep(). Hence I have removed sleep from code for now.

Reply:
ack